### PR TITLE
feat: login user,login page, user context

### LINKED ISSUE #10: User Auth: Login Page
- add a login page and route 
- submitting it should verify the logged in data from backend and give me the user credentials.
- after successful login i should save the recieved token in localstorage
- and redirect page to `/products` and cleaar the history.

### PR DESCRIPTION
**Pull request by tushartiwari7**
# PR for #10 

- added login page. 
- submitting it should verify the logged in data from backend and give me the user credentials.
- after successful login i am saving the recieved token in localstorage
- and redirecting page to `/products` and clearing the history.

localhost_3000_login (1)

### CODE REVIEW

File: src/context/user/UserContext.jsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { createContext } from "react";
+import { useState,useContext } from "react";
+
+const UserContext = createContext();
+
+export const UserProvider = ({children}) => {
+
+  const [user,setUser] = useState({isLoggedIn: false});
```
Comment (JantuDeb) [inline src/context/user/UserContext.jsx:8]: You might need `userName` or `userId` in future. So,  you can add the user to the state variable.

File: src/services/getUser.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import axios from 'axios';
+
+export const  getUser = async (email,password) => {
+  try {
+    const {data,status} = await axios.post("/api/auth/login",{email,password});
+    return {data,status};
+  } catch (error) {
+    console.error("LOGGING IN FAILED",error);
+  }
```
Comment (JantuDeb) [inline src/services/getUser.js]: In case of `error`, you can return it as well

File: src/pages/Login/Login.jsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  const { setUser } = useUser();
+  const navigator = useNavigate();
+
+  const formHandler = (e) => {
+    e.preventDefault();
+    const [email, password] = e.target.elements;
+
+    (async () => {
+      try {
+        const { data, status } = await getUser(email.value, password.value);
```
Comment (JantuDeb) [inline src/pages/Login/Login.jsx]: You don't need to `await` it again as `getUser ` will give resolved promise.

File: src/pages/Login/Login.jsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        if (status === 200) {
+          localStorage.setItem("token", JSON.stringify(data.encodedToken));
+          setUser({ ...data.foundUser, isLoggedIn: true });
+          navigator("/products", { replace: true });
+        } else console.error("EMAIL OR PASSWORD IS INCORRECT", data);
+      } catch (err) {
+        console.log("FAILED TO UPDATE USER", err);
+      }
```
Comment (JantuDeb) [inline src/pages/Login/Login.jsx]: You have added a `try{} catch(e){} ` in `getUser()`  so you can remove it from here

**Review by JantuDeb — APPROVED** [review]
I have pointed out some minor issues.

File: src/context/user/UserContext.jsx (hunk 1 above)
Comment (tushartiwari7) [reply src/context/user/UserContext.jsx:8]: After the user gets loggedin my `user` state will look like this.
and `_id` and `name` are accessible from it.
```json
{
  "_id": "15adeed5-070a-4ffd-828a-11c86a110d56",
  "firstName": "Adarsh",
  "lastName": "Balika",
  "email": "adarshbalika@gmail.com",
  "createdAt": "2022-03-19T11:29:18+05:30",
  "updatedAt": "2022-03-19T11:29:18+05:30",
  "cart": "[]",
  "wishlist": "[]",
  "id": "1",
  "isLoggedIn": true
}
```

File: src/pages/Login/Login.jsx (hunk 3 above)
Comment (tushartiwari7) [reply src/pages/Login/Login.jsx]: Please correct me if i'm wrong but `getUser` is an `async` function so they should return a promise and to capture its returned data we have to use `await`.

File: src/pages/Login/Login.jsx (hunk 4 above)
Comment (tushartiwari7) [reply src/pages/Login/Login.jsx]: Done.

File: src/services/getUser.js (hunk 2 above)
Comment (tushartiwari7) [reply src/services/getUser.js]: done

**Review by Bhakti199 — APPROVED** [review]
- This code looks fine to me.
- Proper hierarchy is maintained.
- Import export from common index file is also a good idea.
- Overall code is well structured.
- This PR is good to merge.

**Pull request merged**